// @lianjia/zola-core/build/index.js
'use strict';

var React = require('react');
var ReactDOM = require('react-dom');
var ReactRouter = require('react-router');

var Page = require('./components/Page');
var Base = require('./components/Base');
var Module = require('./components/Module');

var api = require('./api');
var render = require('./render');
var config = require('./config');
var router = require('./router');

var utils = {
  ajax: require('./utils/ajax'),
  appendSearch: require('./utils/appendSearch'),
  assert: require('./utils/assert'),
  asyncComponent: require('./utils/asyncComponent'),
  deepMerge: require('./utils/deepMerge'),
  getType: require('./utils/getType'),
  inherit: require('./utils/inherit'),
  jsonp: require('./utils/jsonp'),
  noop: require('./utils/noop'),
  object2Query: require('./utils/object2Query'),
  object2Search: require('./utils/object2Search'),
  objectForEach: require('./utils/objectForEach'),
  objectWith: require('./utils/objectWith'),
  objectWithout: require('./utils/objectWithout'),
  query2Object: require('./utils/query2Object'),
  string2Array: require('./utils/string2Array'),
  subscribe: require('./utils/subscribe')
};

var Subscribe = require('./utils/subscribe');
var sub = new Subscribe();

var zola = {};
Object.assign(zola, config, ReactRouter, {
  React: React,
  ReactDOM: ReactDOM,
  Subscribe: Subscribe,
  Base: Base,
  Page: Page,
  Module: Module,
  api: api,
  router: router,
  render: render,
  utils: utils,
  on: sub.on.bind(sub),
  off: sub.off.bind(sub),
  emit: sub.emit.bind(sub)
});

module.exports = Object.assign(zola, { default: zola });
