// @lianjia/zola-core/lib/index.js
const React = require('react')
const ReactDOM = require('react-dom')
const ReactRouter = require('react-router')

const Page = require('./components/Page')
const Base = require('./components/Base')
const Module = require('./components/Module')

const api = require('./api')
const render = require('./render')
const config = require('./config')
const router = require('./router')

const utils = {
  ajax: require('./utils/ajax'),
  appendSearch: require('./utils/appendSearch'),
  assert: require('./utils/assert'),
  asyncComponent: require('./utils/asyncComponent'),
  deepMerge: require('./utils/deepMerge'),
  getType: require('./utils/getType'),
  inherit: require('./utils/inherit'),
  jsonp: require('./utils/jsonp'),
  noop: require('./utils/noop'),
  object2Query: require('./utils/object2Query'),
  object2Search: require('./utils/object2Search'),
  objectForEach: require('./utils/objectForEach'),
  objectWith: require('./utils/objectWith'),
  objectWithout: require('./utils/objectWithout'),
  query2Object: require('./utils/query2Object'),
  string2Array: require('./utils/string2Array'),
  subscribe: require('./utils/subscribe'),
}

const Subscribe = require('./utils/subscribe')
const sub = new Subscribe()

const zola = {}
Object.assign(zola, config, ReactRouter, {
  React,
  ReactDOM,
  Subscribe,
  Base,
  Page,
  Module,
  api,
  router,
  render,
  utils,
  on: sub.on.bind(sub),
  off: sub.off.bind(sub),
  emit: sub.emit.bind(sub),
})

module.exports = Object.assign(zola, { default: zola })
